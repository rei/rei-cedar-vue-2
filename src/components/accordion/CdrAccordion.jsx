import clsx from 'clsx';
import IconCaretDown from '../icon/comps/caret-down';
import modifier from '../../mixins/modifier';
import style from './styles/CdrAccordion.scss';

export default {
  name: 'CdrAccordion',
  components: {
    IconCaretDown,
  },
  mixins: [modifier],
  props: {
    /**
     * The unique id of an accordion.
     */
    id: {
      type: String,
      required: true,
    },
    /**
     * The readable text on the accordion button.
     */
    label: {
      type: String,
      required: false,
    },
    /**
     * Toggle this value to open/close the accordion.
     */
    opened: {
      type: Boolean,
      default: false,
    },
    /**
     * Sets a compact style.
     */
    compact: {
      type: Boolean,
      default: false,
    },
    /**
     * Sets a border-aligned style.
     */
    borderAligned: {
      type: Boolean,
      default: false,
    },
    // Sets the heading level
    level: {
      type: [String, Number],
      required: true,
    },
  },
  data() {
    return {
      focused: false,
      maxHeight: this.opened ? 'none' : 0, // set maxHeight to none if initialized as open
      style,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-accordion';
    },
    styleClass() {
      const styles = [];

      if (this.compact) {
        styles.push(this.modifyClassName(this.baseClass, 'compact'));
      }

      if (this.borderAligned) {
        styles.push(this.modifyClassName(this.baseClass, 'border-aligned'));
      }

      return styles.join(' ');
    },
    focusedClass() {
      return this.focused ? this.modifyClassName(this.baseClass, 'focused') : null;
    },
    isOpenClass() {
      return this.opened ? 'open' : 'closed';
    },
  },
  watch: {
    opened() {
      this.maxHeight = this.opened ? `${this.$refs['accordion-content'].clientHeight}px` : 0;
    },
  },
  mounted() {
    /*
      The intent here is to give maxHeight an actual pixel value when the accordion
      is initialized as open. This guarantees that the open-to-close animation is
      nice and smooth the first time they click it.
    */
    if (this.opened && this.$refs['accordion-content']) {
      this.maxHeight = `${this.$refs['accordion-content'].clientHeight}px`;
    }
  },
  methods: {
    onClick(event) {
      this.$emit('accordion-toggle', event);
    },
    onFocus() {
      this.focused = true;
    },
    onBlur() {
      this.focused = false;
    },
  },
  render() {
    const Heading = `h${this.level}`;

    return (<div
      class={clsx(this.style[this.baseClass],
        this.modifierClass,
        this.styleClass,
        this.focusedClass)}
      id={`${this.id}-accordion`}
      ref="accordion-container"
    >
      <Heading class={this.style['cdr-accordion__header']}>
        <button
          class={[this.style['cdr-accordion__button'], 'js-cdr-accordion-button']} // .js-accordion-button is for group focus management
          id={this.id}
          onClick={this.onClick}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          aria-expanded={`${this.opened}`}
          aria-controls={`${this.id}-collapsible`}
          >
          <span
            class={this.style['cdr-accordion__label']}
            id={`${this.id}-label`}
            >
            { this.$slots.label || this.label}
          </span>
          <icon-caret-down
            class={clsx(this.style['cdr-accordion__icon'], this.isOpenClass)}
            size={this.compact ? 'small' : null}
            />
        </button>
      </Heading>
      <div
        class={clsx(this.style['cdr-accordion__content-container'], this.isOpenClass)}
        style={ { maxHeight: this.maxHeight } }
      >
        <div
          class={clsx(this.style['cdr-accordion__content'], this.isOpenClass)}
          aria-hidden={`${!this.opened}`}
          id={`${this.id}-collapsible`}
          ref="accordion-content"
        >
          {this.$slots.default}
        </div>
      </div>
    </div>);
  },
};
