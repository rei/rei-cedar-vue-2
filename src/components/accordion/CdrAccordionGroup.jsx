import clsx from 'clsx';
import style from './styles/CdrAccordionGroup.scss';

export default {
  name: 'CdrAccordionGroup',
  data() {
    return {
      style,
      accordionButtons: [],
      currentIdx: 0,
    };
  },
  computed: {
    nextIdx() {
      const idx = this.currentIdx + 1;
      return idx >= this.accordionButtons.length ? 0 : idx; // if at the last, go to first
    },
    prevIdx() {
      const idx = this.currentIdx - 1;
      return idx <= -1 ? (this.accordionButtons.length - 1) : idx; // if at first, go to last
    },
  },
  mounted() {
    // get all of the buttons in the group
    this.accordionButtons = this.$el.querySelectorAll('.js-cdr-accordion-button');
  },
  methods: {
    handleKeyDown(e) {
      // something besides the button is focused
      if (this.currentIdx === -1) return;

      const { key } = e;
      switch (key) {
        case 'Home':
          e.preventDefault();
          this.accordionButtons[0].focus();
          break;
        case 'End':
          e.preventDefault();
          this.accordionButtons[this.accordionButtons.length - 1].focus();
          break;
        case 'ArrowDown':
        case 'Down':
          e.preventDefault();
          this.accordionButtons[this.nextIdx].focus();
          break;
        case 'ArrowUp':
        case 'Up':
          e.preventDefault();
          this.accordionButtons[this.prevIdx].focus();
          break;
        default: break;
      }
    },
    focusin(e) {
      // find out which, if any, button is focused
      this.currentIdx = Array.prototype.indexOf.call(this.accordionButtons, e.target);
    },
  },
  render(h) {
    return (<ul
      class={clsx(this.style['cdr-accordion-group'])}
      onFocusin={this.focusin}
      onKeydown={this.handleKeyDown}
    >
        {
          this.$slots.default.map(vNode => (vNode.tag !== undefined ? h('li', vNode, [vNode]) : ''))
        }
    </ul>);
  },
};
