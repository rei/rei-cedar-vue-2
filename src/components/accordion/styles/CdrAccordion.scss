@import '../../../css/settings/index.scss';
@import url('@rei/cedar/dist/style/cdr-icon.css');

$accordion-border: 1px solid $cdr-color-border-primary;

.cdr-accordion {
  border-top: $accordion-border;
  border-bottom: $accordion-border;
  position: relative;
  z-index: 1;

  &+& {
    border-top: none;
  }

  &__button {
    @include cdr-text-utility-sans-strong-300;

    background-color: transparent;
    border: none;
    color: $cdr-color-text-primary;
    cursor: pointer;
    display: flex;
    justify-content: space-between;
    padding: $cdr-space-inset-one-x;
    position: relative;
    text-align: left;
    width: 100%;
    z-index: 1;

    &:hover,
    &:focus {
      background-color: $cdr-color-background-accordion-hover;
    }
  }

  &__label {
    cursor: pointer;
    margin-bottom: 0;
  }

  &__header {
    margin: 0;
    padding: 0;
    font-weight: unset;
    font-style: unset;
    font-size: unset;
  }

  &__icon {
    margin-left: $cdr-space-one-x;
    fill: $cdr-color-icon-default;
    transition: transform $cdr-duration-3-x $cdr-timing-function-ease-out;

    &:global(.open) {
      transform: rotate(-180deg);
    }
  }

  &__content-container {
    overflow: hidden;
    transition-duration: $cdr-duration-3-x;
    transition-property: max-height;
    transition-timing-function: cubic-bezier(0.15, 0, 0.15, 1); // TODO: possible token?

    &:global(.open) {
      transition-duration: $cdr-duration-3-x;
    }
  }

  &__content {
    @include cdr-text-body-300;

    background-color: transparent;
    color: $cdr-color-text-primary;
    padding: $cdr-space-half-x $cdr-space-one-x $cdr-space-one-x;
    transition: opacity $cdr-duration-1-x linear;
    opacity: 0;

    & > * {
      @include cdr-text-body-300;
    }

    &:global(.open) {
      opacity: 1;
    }

    &:global(.closed) {
      visibility: hidden;
    }
  }

  /* Style variants
    ========================================================================== */

  /* Border-Aligned
     ========== */
  &--border-aligned {
    margin: 0 $cdr-space-inset-one-x;

    .cdr-accordion__button {
      margin-left: -$cdr-space-one-x;
      width: calc(100% + (#{$cdr-space-one-x} * 2));
    }

    .cdr-accordion__content {
      padding: $cdr-space-half-x 0 $cdr-space-one-x;
    }
  }

  /* Compact
     ========== */
  &--compact {
    .cdr-accordion__button {
      @include cdr-text-utility-sans-strong-200;

      padding: $cdr-space-inset-three-quarter-x;
    }

    .cdr-accordion__content {
      @include cdr-text-utility-sans-200;

      padding: $cdr-space-quarter-x $cdr-space-three-quarter-x $cdr-space-three-quarter-x;

      & > * {
        @include cdr-text-utility-sans-200;
      }
    }
  }

  /* Border-Aligned and Compact
     ========== */
  &--compact.cdr-accordion--border-aligned {
    .cdr-accordion__button {
      margin-left: -$cdr-space-three-quarter-x;
      width: calc(100% + (#{$cdr-space-three-quarter-x} * 2));
    }

    .cdr-accordion__content {
      padding: $cdr-space-quarter-x 0 $cdr-space-three-quarter-x;
    }
  }

  &--focused {
    /*
      This z-index change is to fix a Chrome-only visual issue with the light blue focus outline. Basically, the bottom
      line of the outline is usually covered by the accordion immediately below what's currently focuesd. We increase
      the z-index so that the outline can be correctly seen when the accordion is focused.
    */
    z-index: 10;
  }
}
