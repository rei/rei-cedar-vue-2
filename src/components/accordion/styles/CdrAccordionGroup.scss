@import '../../../css/settings/index.scss';

.cdr-accordion-group {
  display: block;
  list-style: none;
  padding: 0;
  margin: 0;

  li + li {
    & > .cdr-accordion {
      border-top: none;
    }
  }

}
