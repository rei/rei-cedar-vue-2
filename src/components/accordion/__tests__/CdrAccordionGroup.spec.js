import { shallowMount, mount } from '@vue/test-utils';
import CdrAccordionGroup from 'componentdir/accordion/CdrAccordionGroup';
import CdrAccordion from '../CdrAccordion';

const propsData = {
  id: 'test',
  label: 'A label',
  compact: false,
  borderAligned: false,
};

describe('CdrAccordionGroup', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(CdrAccordionGroup, {
      stubs: {
        'cdr-accordion': CdrAccordion,
      },
      slots: {
        default: [
          '<cdr-accordion id="tab1" label="label1" level="2"/>',
          '<cdr-accordion id="tab2" label="label2" level="2"/>'
        ]
      },
    });
    expect(wrapper.element).toMatchSnapshot()
  });
  
  it('has correct a11y', () => {
    const wrapper = mount(CdrAccordionGroup, {
      stubs: {
        'cdr-accordion': CdrAccordion,
      },
      slots: {
        default: [
          '<cdr-accordion id="tab1" label="label1" level="2"/>',
          '<cdr-accordion id="tab2" label="label2" level="2"/>',
          '<cdr-accordion id="tab3" label="label3" level="2"/>',
        ]
      },
      attachToDocument: true,
    });

    expect(wrapper.is('ul')).toBe(true);
    /* keyboard nav tests
      `focusin` doesn't fire outside of the browser environment so it's faked by just doing the logic
      manually instead with setData. The proper focus logic is still checked via document.activeElement
    */
    const buttons = wrapper.vm.accordionButtons;
    // Up (first to last)
    wrapper.trigger('keydown', { key: 'ArrowUp' });
    wrapper.setData({ currentIdx: 2 });
    expect(buttons[2]).toBe(document.activeElement);
    // Down (last to first)
    wrapper.trigger('keydown', { key: 'ArrowDown' });
    wrapper.setData({ currentIdx: 0 });
    expect(buttons[0]).toBe(document.activeElement);
    // Down
    wrapper.trigger('keydown', { key: 'ArrowDown' });
    wrapper.setData({ currentIdx: 1 });
    expect(buttons[1]).toBe(document.activeElement);
    // Up
    wrapper.trigger('keydown', { key: 'ArrowUp' });
    wrapper.setData({ currentIdx: 0 });
    expect(buttons[0]).toBe(document.activeElement);
    // End
    wrapper.trigger('keydown', { key: 'End' });
    wrapper.setData({ currentIdx: 2 });
    expect(buttons[2]).toBe(document.activeElement);
    // Home
    wrapper.trigger('keydown', { key: 'Home' });
    wrapper.setData({ currentIdx: 0 });
    expect(buttons[0]).toBe(document.activeElement);
  });
});