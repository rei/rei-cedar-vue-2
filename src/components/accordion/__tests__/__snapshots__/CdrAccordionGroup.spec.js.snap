// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrAccordionGroup renders correctly 1`] = `
<ul
  class="cdr-accordion-group"
>
  <li>
    <div
      class="cdr-accordion"
      id="tab1-accordion"
    >
      <h2
        class="cdr-accordion__header"
      >
        <button
          aria-controls="tab1-collapsible"
          aria-expanded="false"
          class="cdr-accordion__button js-cdr-accordion-button"
          id="tab1"
        >
          <span
            class="cdr-accordion__label"
            id="tab1-label"
          >
            label1
          </span>
          <icon-caret-down-stub
            class="cdr-accordion__icon closed"
            space=""
          />
        </button>
      </h2>
      <div
        class="cdr-accordion__content-container closed"
        style="max-height: 0;"
      >
        <div
          aria-hidden="true"
          class="cdr-accordion__content closed"
          id="tab1-collapsible"
        />
      </div>
    </div>
  </li>
  <li>
    <div
      class="cdr-accordion"
      id="tab2-accordion"
    >
      <h2
        class="cdr-accordion__header"
      >
        <button
          aria-controls="tab2-collapsible"
          aria-expanded="false"
          class="cdr-accordion__button js-cdr-accordion-button"
          id="tab2"
        >
          <span
            class="cdr-accordion__label"
            id="tab2-label"
          >
            label2
          </span>
          <icon-caret-down-stub
            class="cdr-accordion__icon closed"
            space=""
          />
        </button>
      </h2>
      <div
        class="cdr-accordion__content-container closed"
        style="max-height: 0;"
      >
        <div
          aria-hidden="true"
          class="cdr-accordion__content closed"
          id="tab2-collapsible"
        />
      </div>
    </div>
  </li>
</ul>
`;
